File: services/global/util/src/main/java/com/tars/util/common/RefCountHolder.java
package com.tars.util.common;

import java.util.function.Consumer;
import java.util.function.Supplier;

public class RefCountHolder<T> {

  private RefCountToggle toggle;
  private T instance = null;

  public RefCountHolder(Supplier<T> supplier, Consumer<T> finalizer) {
    toggle = new RefCountToggle(
        () -> instance = supplier.get(),
        () -> {
          finalizer.accept(instance);
          instance = null;
        }
    );
  }

  public T requestInstance() {
    toggle.increment();
    return instance;
  }

  public void returnInstance(T instance) {
    if (instance == this.instance) {
      toggle.decrement();
    } else {
      throw new IllegalStateException("resource does not belong to holder");
    }
  }
}

File: sources/global/util/src/main/kotlin/util/misc/RefCountHolder.kt
package util.misc

import util.global.areSame
import util.global.condition
import util.global.getMandatory

class RefCountHolder<T>(val supplier: () -> T, val finalizer: (T) -> Unit) {

    private val toggle: RefCountToggle
    private var instance: T? = null

    init {
        toggle = RefCountToggle(
                { instance = supplier.invoke() },
                {
                    finalizer.invoke(getMandatory(instance))
                    instance = null
                }
        )
    }

    fun requestInstance(): T {
        toggle.increment()
        return getMandatory(instance)
    }

    fun returnInstance(instance: T?) {
        condition(areSame(this.instance, instance))
        toggle.decrement()
    }
}
